### PR TITLE
[google-cloud-cpp] add opentelemetry feature

### PR DESCRIPTION
**Pull request by dbolduc**
Tested locally (on x64-linux) with:

```sh
for feature in storage experimental-opentelemetry; do ./vcpkg remove google-cloud-cpp; ./vcpkg install "google-cloud-cpp[core,${feature}]" || break; done
```

and

```sh
./vcpkg remove google-cloud-cpp && ./vcpkg install 'google-cloud-cpp[*]'
```

cc @coryan

### CODE REVIEW

**Comment by dbolduc** [comment]
@microsoft-github-policy-service agree company="Google"

File: ports/google-cloud-cpp/vcpkg.json [hunk 1]
```diff
@@ -8,6 +8,10 @@
   "dependencies": [
     "abseil",
     "grpc",
+    {
+      "name": "opentelemetry-cpp",
+      "host": true
```
Comment (coryan) [inline ports/google-cloud-cpp/vcpkg.json:13]: I do not think this is true. `"host: true` implies this dep is needed on the host platform during a cross-compilation build.  And you really need the *target* platform.

File: ports/google-cloud-cpp/vcpkg.json [hunk 2]
```diff
@@ -8,6 +8,10 @@
   "dependencies": [
     "abseil",
     "grpc",
+    {
```
Comment (coryan) [inline ports/google-cloud-cpp/vcpkg.json:11]: You are missing the `port-version` number above:

https://learn.microsoft.com/en-us/vcpkg/users/versioning#port-version

This is omitted when it is `0`, but here you need to set it to `1`.

File: ports/google-cloud-cpp/vcpkg.json [hunk 3]
```diff
@@ -828,6 +832,25 @@
         }
       ]
     },
+    "opentelemetry": {
+      "description": "OpenTelemetry C++ GCP Exporter Library",
+      "dependencies": [
+        {
```
Comment (coryan) [inline ports/google-cloud-cpp/vcpkg.json:838]: We may want a `rest-common` dependency (analogous to `grpc-common`) to refactor the deps for this feature and for `storage`.

File: ports/google-cloud-cpp/vcpkg.json [hunk 4]
```diff
@@ -828,6 +832,25 @@
         }
       ]
     },
+    "opentelemetry": {
```
Comment (coryan) [inline ports/google-cloud-cpp/vcpkg.json:835]: Why not name this `experimental-opentelemetry`?  It is experimental, right?  We can always keep the `experimental-*` shim even after it reaches GA.

**Review by coryan — CHANGES_REQUESTED** [review]

**Comment by coryan** [comment]
You may want to manually test using `vcpkg install google-cloud-cpp[*]` also.

File: ports/google-cloud-cpp/vcpkg.json (hunk 1 above)
Comment (dbolduc) [reply ports/google-cloud-cpp/vcpkg.json]: Fixed, thanks.

File: ports/google-cloud-cpp/vcpkg.json (hunk 2 above)
Comment (dbolduc) [reply ports/google-cloud-cpp/vcpkg.json]: I forgot this. I need to add another commit.

Comment (dbolduc) [reply ports/google-cloud-cpp/vcpkg.json]: Fixed, thanks.

File: ports/google-cloud-cpp/vcpkg.json (hunk 3 above)
Comment (dbolduc) [reply ports/google-cloud-cpp/vcpkg.json]: Done. Good idea, given `compute`, `sql`.

File: ports/google-cloud-cpp/vcpkg.json (hunk 4 above)
Comment (dbolduc) [reply ports/google-cloud-cpp/vcpkg.json]: You're right. And I now see that is what we do for `experimental-storage-grpc`

File: ports/google-cloud-cpp/vcpkg.json [hunk 5]
```diff
@@ -1,13 +1,15 @@
...
   "version": "2.13.0",
+  "port-version": 1,
   "description": "C++ Client Libraries for Google Cloud Platform APIs.",
   "homepage": "https://github.com/googleapis/google-cloud-cpp",
   "license": "Apache-2.0",
   "supports": "!uwp",
   "dependencies": [
     "abseil",
     "grpc",
+    "opentelemetry-cpp",
```
Comment (coryan) [inline ports/google-cloud-cpp/vcpkg.json:12]: I think this dependency is only required if `experimental-opentelemetry` is enabled.  I think it belongs in the dependencies of that feature.

**Review by coryan — APPROVED** [review]
A couple of small changes and you should make this into a non-draft PR.

My approval is not that useful. <insert "no power here meme" or something>

File: ports/google-cloud-cpp/vcpkg.json (hunk 5 above)
Comment (dbolduc) [reply ports/google-cloud-cpp/vcpkg.json]: Right. Fixed.

I confused myself. I was thinking, "I always want the `opentelemetry-cpp` library available" but I forgot that vcpkg is the one compiling our code (i.e. defining `-DGOOGLE_CLOUD_CPP_ENABLE=...`).

So maybe what I want is `experimental-opentelemetry` as a `default_feature` but that is a separate decision, which I will defer.

File: versions/g-/google-cloud-cpp.json [hunk 6]
```diff
@@ -1,7 +1,12 @@
 {
   "versions": [
     {
-      "git-tree": "c067a236b62d38d1f56a658b086039bf22441c4a",
+      "git-tree": "d9ee54a1cac40ce75bb8684d4e8bd146fe90f214",
+      "version": "2.13.0",
+      "port-version": 1
+    },
+    {
+      "git-tree": "a14e88242b5cc905b0620d3c1a149c9277842542",
```
[the remark this answers is not in the document]
Comment (dbolduc) [reply versions/g-/google-cloud-cpp.json]: Fixed, thanks.

That must have been an artifact from failing to update the `port-version` in previous commits.

**Comment by Adela0814** [comment]
I encountered the following error while testing the feature:
Repro step:
`.\vcpkg.exe install google-cloud-cpp[*]:x64-windows`
Error:
`E:\1234\vcpkg\buildtrees\google-cloud-cpp\src\v2.13.0-c4319f8bec.clean\google\cloud\vmmigration\v1\internal\vm_migration_connection_impl.cc : fatal error C1128: number of sections exceeded object file format limit: compile with /bigobj`
[install-x64-windows-dbg-out.log](https://github.com/microsoft/vcpkg/files/12211894/install-x64-windows-dbg-out.log)

**Comment by Adela0814** [comment]
Note: I will be converting your PR to draft status. When you respond, please revert to "ready for review". That way, I can be aware that you've responded since you can't modify the tags.

**Comment by dbolduc** [comment]
> I encountered the following error while testing the feature: Repro step: `.\vcpkg.exe install google-cloud-cpp[*]:x64-windows` Error: `E:\1234\vcpkg\buildtrees\google-cloud-cpp\src\v2.13.0-c4319f8bec.clean\google\cloud\vmmigration\v1\internal\vm_migration_connection_impl.cc : fatal error C1128: number of sections exceeded object file format limit: compile with /bigobj` [install-x64-windows-dbg-out.log](https://github.com/microsoft/vcpkg/files/12211894/install-x64-windows-dbg-out.log)

Thanks for looking. The `/bigobj` error is hopefully fixed upstream by https://github.com/googleapis/google-cloud-cpp/pull/12272

I am going to close this PR. We will add the `experimental-opentelemetry` feature in our next release of the library (`v2.14.0#0`, scheduled for tomorrow).

**Pull request closed without merging**